### PR TITLE
sds: clusters and listeners read static secrets from Bootstrap.static_resources

### LINKED ISSUE #1194: Secret Discovery Service for v2 APIs
Hi,

We want to be able to rotate Envoy certs/keys without having to restart the Envoy.  We're particularly interested in the client cert/key for connecting to upstream clusters.  We recognize the CDS allows us to dynamically change clusters, but that doesn't let us rotate the cert used to connect to the CDS itself.

We see prior discussion about supporting cert rotation (https://github.com/lyft/envoy/issues/891) which was answered by pointing people towards the in-development LDS (https://github.com/lyft/envoy/issues/315).  However, we don't see how the LDS solves the problem of rotating certs used to connect to the CDS.

What work would be required for support cert/key rotation for CDS without having to restart Envoy?  This issue is a blocker for us using Envoy (we're looking at integrating it, and perhaps Istio, into Cloud Foundry).

We might be able to contribute the feature, if we had some guidance.

Cheers,
Angela and @rosenhouse

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
*Description*:
Clusters and listeners read secrets from the static resources in the bootstrap configuration.
Reading secrets from the Secret Discovery Service will follow.

Risk Level: Low

Fixes #1194 

Signed-off-by: Jae Kim jaebong.kim@gmail.com

### CODE REVIEW

**Comment by mattklein123** [comment]
Excited to see this being worked on. @lizan @PiotrSikora can you take a first pass?

File: ci/do_circle_ci.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ set -e
 
 # bazel uses jgit internally and the default circle-ci .gitconfig says to
 # convert https://github.com to ssh://git@github.com, which jgit does not support.
-mv ~/.gitconfig ~/.gitconfig_save
+# mv ~/.gitconfig ~/.gitconfig_save
```
Comment (lizan) [inline ci/do_circle_ci.sh]: ?

File: include/envoy/secret/secret.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   * @return a string of certificate chain
+   */
+  virtual const std::string& getCertificateChain() PURE;
+  /**
+   * @return a string of private key
+   */
+  virtual const std::string& getPrivateKey() PURE;
+};
+
+typedef std::shared_ptr<Secret> SecretPtr;
```
Comment (lizan) [inline include/envoy/secret/secret.h]: nit: `SecretSharedPtr`

File: source/common/json/config_schemas.cc [hunk 3]
```diff
@@ -1678,4 +1678,19 @@ const std::string Json::Schema::SDS_SCHEMA(R"EOF(
     "required" : ["hosts"]
   }
   )EOF");
+
```
Comment (lizan) [inline source/common/json/config_schemas.cc]: This is not going to v1 so I don't think you need this.

File: source/common/secret/secret_impl.cc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace Envoy {
+namespace Secret {
+
+SecretImpl::SecretImpl(const envoy::api::v2::auth::Secret& config)
+    : name_(config.name()),
+      certificate_chain_(readDataSource(config.tls_certificate().certificate_chain(), true)),
+      private_key_(readDataSource(config.tls_certificate().private_key(), true)) {}
+
+const std::string SecretImpl::readDataSource(const envoy::api::v2::core::DataSource& source,
```
Comment (lizan) [inline source/common/secret/secret_impl.cc]: Just use `Config::DataSource::read`

File: source/common/secret/secret_impl.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include "envoy/secret/secret.h"
+
+namespace Envoy {
+namespace Secret {
+
+class SecretImpl : public Secret {
+public:
+  SecretImpl(const envoy::api::v2::auth::Secret& config);
+
+  virtual ~SecretImpl() {}
```
Comment (lizan) [inline source/common/secret/secret_impl.h]: no need of this line

File: source/common/secret/secret_impl.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  const std::string& getName() override { return name_; }
+
+  const std::string& getCertificateChain() override { return certificate_chain_; }
+
+  const std::string& getPrivateKey() override { return private_key_; }
+
+private:
+  const std::string readDataSource(const envoy::api::v2::core::DataSource& source,
+                                   bool allow_empty);
+  const std::string getDataSourcePath(const envoy::api::v2::core::DataSource& source);
```
Comment (lizan) [inline source/common/secret/secret_impl.h]: Unused?

Comment (mangchiandjjoe) [reply source/common/secret/secret_impl.h]: Removed

File: include/envoy/secret/secret.h (hunk 2 above)
Comment (mangchiandjjoe) [reply include/envoy/secret/secret.h]: Replaced

File: include/envoy/secret/secret.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * Secret contains certificate chain and private key
+ */
+class Secret {
+public:
+  virtual ~Secret() {}
+
+  /**
+   * @return a name of the SDS secret
+   */
+  virtual const std::string& getName() PURE;
```
Comment (lizan) [inline include/envoy/secret/secret.h]: make these getters as const method?

Also I would prefer not having `get` prefix for those properties, so just `name`, `certificateChain`...

File: source/common/secret/secret_manager_impl.cc [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+namespace Envoy {
+namespace Secret {
+
+bool SecretManagerImpl::addOrUpdateStaticSecret(const SecretSharedPtr secret) {
+  static_secrets_[secret->getName()] = secret;
+  return true;
+}
+
+SecretSharedPtr SecretManagerImpl::getStaticSecret(const std::string& name) {
+  return (static_secrets_.find(name) != static_secrets_.end()) ? static_secrets_[name] : nullptr;
```
Comment (lizan) [inline source/common/secret/secret_manager_impl.cc]: You're looking up twice, prefer:

```
auto it = static_secrets_.find(name);
return it != static_secrets_.end() ? it->second : nullptr;
```

File: source/common/secret/secret_manager_impl.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <shared_mutex>
```
Comment (lizan) [inline source/common/secret/secret_manager_impl.h]: unused

File: include/envoy/secret/secret.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  virtual const std::string& getCertificateChain() PURE;
+  /**
+   * @return a string of private key
+   */
+  virtual const std::string& getPrivateKey() PURE;
+};
+
+typedef std::shared_ptr<Secret> SecretSharedPtr;
+
+typedef std::unordered_map<std::string, SecretSharedPtr> SecretSharedPtrMap;
```
Comment (lizan) [inline include/envoy/secret/secret.h]: The map and vector are not used in interfaces, so move them to impl.h?

Comment (mangchiandjjoe) [reply include/envoy/secret/secret.h]: Moved,

**Comment by lizan** [comment]
@mangchiandjjoe Can you try merge master?

File: include/envoy/secret/secret.h (hunk 7 above)
Comment (mangchiandjjoe) [reply include/envoy/secret/secret.h]: Done

**Comment by lizan** [comment]
FIrst pass done, @mattklein123 can you take a look?

**Comment by mattklein123** [comment]
@mangchiandjjoe since we are going to have to fix DCO eventually, can you potentially just do it now before we start more reviews? Feel free to squash/rebase/force push.

**Comment by mattklein123** [comment]
Presumably https://github.com/envoyproxy/envoy/pull/3465 replaces this so closing.

**Pull request closed without merging**